### DISCUSSION TITLE
IV curve measurement with Keithley 2470!

### DISCUSSION THREAD

**Discussion by fgarcia01** [pymeasure/pymeasure#576] [Q&A]
Dear Developers,
Can somebody help to get the IV curve measurement with a Keithley 2470? There are scripts around but they do only the Current sweep in one module and measure Voltage in another, neither both are 2470.
Many thanks in advance!

**Comment by neal-kepler** [comment]
I quickly took a look at the 2470 documentation and it looks like it may accept the 2400 command set in addition to the newer tsp lua style commands. So provided you have a driver that works, the easiest thing would be to tell the 2470 to also measure_voltage() (if you want to source current still and not the reverse) at like line 51 of the [ IV example](https://github.com/pymeasure/pymeasure/blob/master/examples/Current-Voltage%20Measurements/iv_keithley.py)
```
self.source.measure_voltage()
```
and then replace line 70 with
```
voltage = self.source.voltage
```
i.e. just get the voltage from the Keithley 2470 itself. Then you can just cut out all the parts relating to the other meter.

**Reply by fgarcia01** [reply]
Thanks for this script. I tested and the communications work nicely (when changing these lines: pymeasure.instruments.keithley import Keithley2450, self.meter = Keithley2450("GPIB::18") and self.source = Keithley2450("GPIB::18")). 
However, this script works like this: set current between two values and increment a step, and measure voltage. 
In my case, I would need just the opposite to set voltage and read current and on this, I am getting stuck.

**Reply by neal-kepler** [reply]
Great, if the driver works for the set I measure V case, then it should also work for the set V measure I case. It _should_ be as simple as changing every instance of "current" to "voltage" (i.e. ```self.source.apply_current()``` to ```self.source.apply_voltage()``` ) and updating the inputs/compliance/range to be sensible for set of voltages. If something breaks the driver documentation is pretty good.

**Reply by fgarcia01** [reply]
I continue fighting with this script and just chaning current by voltage and viceversa do not make the job. Would you like I place the script?

**Reply by neal-kepler** [reply]
Yeah, I'm happy to take a look

**Reply by fgarcia01** [reply]
I think this script now it works, but I suspect it is trying to set voltage and read current from the rear and not the front. Do you know how to force this?
Is ok, to post the script here?

**Reply by bilderbuchi** [reply]
Yes, post it here. Surround it with lines with triple backticks(```) to get readable code formatting.
As to your question, the 2400/2450 have a method `use_front_terminals()`: https://pymeasure.readthedocs.io/en/latest/api/instruments/keithley/keithley2450.html?highlight=front#pymeasure.instruments.keithley.Keithley2450.use_front_terminals

**Reply by fgarcia01** [reply]
""""
This example demonstrates how to make a graphical interface to preform
IV characteristic measurements. There are a two items that need to be 
changed for your system:

1) Correct the GPIB addresses in IVProcedure.startup for your instruments
2) Correct the directory to save files in MainWindow.queue

Run the program by changing to the directory containing this file and calling:

python iv_keithley.py

From original owner this script was modified to set voltage and read current. In addition, it was tested with keithley 2470.

"""




import logging
log = logging.getLogger('')
log.addHandler(logging.NullHandler())

import sys
from time import sleep
import numpy as np

from pymeasure.instruments.keithley import Keithley2450
from pymeasure.instruments import Instrument
from pymeasure.log import console_log
from pymeasure.display.Qt import QtGui
from pymeasure.display.windows import ManagedWindow
from pymeasure.experiment import (
    Procedure, FloatParameter, unique_filename, Results
)


class IVProcedure(Procedure):

    max_voltage = FloatParameter('Maximum Voltage', units='V', default=10)
    min_voltage = FloatParameter('Minimum Voltage', units='V', default=0)
    voltage_step = FloatParameter('Voltage Step', units='V', default=1)
    delay = FloatParameter('Delay Time', units='ms', default=20)
    voltage_range = FloatParameter('Voltage Range', units='V', default=10)

    DATA_COLUMNS = ['Voltage (V)', 'Current (A)', 'Resistance (Ohm)']

    def startup(self):
        log.info("Setting up instruments")
        self.meter = Keithley2450("GPIB::18") # Keithley 2470
        self.meter.measure_current()
        self.meter.voltage_range = self.voltage_range
        self.meter.voltage_nplc = 10 # Integration constant to Medium
        
        self.source = Keithley2450("GPIB::18")
        self.source.apply_voltage()
        self.source.source_current_range = 1e-3 # A
        self.source.complinance_voltage = self.voltage_range
        self.source.enable_source()
        sleep(2)

    def execute(self):
        voltages = np.arange(self.min_voltage, self.max_voltage, self.voltage_step)
        steps = len(voltages)
        
        log.info("Starting to sweep through voltage")
        for i, voltage in enumerate(voltages):
            log.debug("Sweep voltage: %g V" % voltage)

            self.source.source_voltage = voltage
            # Or use self.source.ramp_to_current(current, delay=0.1)
            sleep(self.delay*1e-3)
            
            current = self.meter.current

            if abs(current) <= 1e-10:
                resistance = np.nan
            else:
                resistance = voltage/current
            data = {
                'Voltage (V)': voltage,
                'Current (A)': current,
                'Resistance (Ohm)': resistance
            }
            self.emit('results', data)
            self.emit('progress', 100.*i/steps)
            if self.should_stop():
                log.warning("Catch stop command in procedure")
                break

    def shutdown(self):
        self.source.shutdown()
        log.info("Finished")


class MainWindow(ManagedWindow):

    def __init__(self):
        super(MainWindow, self).__init__(
            procedure_class=IVProcedure,
            inputs=[
                'max_voltage', 'min_voltage', 'voltage_step',
                'delay', 'voltage_range'
            ],
            displays=[
                'max_voltage', 'min_voltage', 'voltage_step',
                'delay', 'voltage_range'
            ],
            x_axis='Voltage (V)',
            y_axis='Current (A)'
        )
        self.setWindowTitle('IV Measurement')

    def queue(self):
        directory = "./" # Change this to the desired directory
        filename = unique_filename(directory, prefix='IV')

        procedure = self.make_procedure()
        results = Results(procedure, filename)
        experiment = self.new_experiment(results)

        self.manager.queue(experiment)


if __name__ == "__main__":
    app = QtGui.QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec_())"

**Reply by bilderbuchi** [reply]
`self.source.complinance_voltage` is a typo - `compliance_voltage`.

**Reply by fgarcia01** [reply]
Ok! now it works. Of course if not perfect but does the job.
Thanks for this support and if can add Keithley 2470 to the library of modules will help some people in the future.
